src/HDF5.jl

```diff
@@ -1370,12 +1378,50 @@ Base.getindex(parent::Union{File,Group}, r::Reference) = _deref(parent, r)
 Base.getindex(parent::Dataset, r::Reference) = _deref(parent, r) # defined separately to resolve ambiguity
 
 # convert special types to native julia types
-normalize_types(x) = x
-normalize_types(x::NamedTuple{T}) where {T} = NamedTuple{T}(map(normalize_types, values(x)))
-normalize_types(x::Cstring) = unsafe_string(x)
-normalize_types(x::FixedString) = unpad(String(collect(x.data)), pad(x))
-normalize_types(x::FixedArray) = normalize_types.(reshape(collect(x.data), size(x)...))
-normalize_types(x::VariableArray) = normalize_types.(copy(unsafe_wrap(Array, convert(Ptr{eltype(x)}, x.p), x.len, own=false)))
+function normalize_types(::Type{T}, buf::AbstractMatrix{UInt8}) where {T}
+    return [_normalize_types(T, view(buf, :, ind)) for ind in axes(buf, 2)]
+end
+
+# high-level description which should always work
+function _typed_load(::Type{T}, buf::AbstractVector{UInt8}) where {T}
+    return @inbounds reinterpret(T, buf)[1]
+end
+# fast-path for common concrete types with simple layout (which should be nearly all
+# cases)
+function _typed_load(::Type{T}, buf::V) where {T, V <: Union{Vector{UInt8}, Base.FastContiguousSubArray{UInt8,1}}}
+    ref = Ref{T}()
+    GC.@preserve buf Base._memcpy!(unsafe_convert(Ptr{Cvoid}, ref), pointer(buf, 1), sizeof(T))
```

Comment:
Maybe just use `pointer(buf)` (it doesn't matter, but I don't see a reason to specifically single out the first element)

Is the `unsafe_convert` to a `Cvoid` needed to get the memory address of `ref`? 

Could you `sptr = pointer(buf)`   `dptr = unsafe_convert(Ptr{Cvoid}, ref)`
on the line above to break this up? 

I did think that `reinterpet` was fast on 1.6, but your benchmarks show that there's still some performance  on the table.

Reply:
> Maybe just use `pointer(buf)` (it doesn't matter, but I don't see a reason to specifically single out the first element)

I thought there was one of the abstract array types which didn't have a bare `pointer` method, but I can't seem to reproduce that anymore. It might have become unnecessary once I figured out the details in https://github.com/JuliaIO/HDF5.jl/pull/745#issuecomment-731631381.

> Is the `unsafe_convert` to a `Cvoid` needed to get the memory address of `ref`?

Yes, since `pointer(::RefValue)` isn't defined. It's what `ccall`s with `Ref`s lower to doing, so I think it's the canonical way to do it.

> Could you `sptr = pointer(buf)` `dptr = unsafe_convert(Ptr{Cvoid}, ref)` on the line above to break this up?

Sure.

Reply:
>> Is the unsafe_convert to a Cvoid needed to get the memory address of ref?
>
> Yes, since pointer(::RefValue) isn't defined. It's what ccalls with Refs lower to doing, so I think it's the canonical way to do it.

Oh, it's also the fastest way to get the correct pointer type. The `memcpy` ccall is defined for `Ptr{UInt8}`, which most `Ref{T}` types won't "decay to" via automatic ccall conversion, but since `Ptr{Cvoid}` is a valid pointer type for any `Ref{T}` and automatically converts to `Ptr{UInt8}` in the ccall, this ends up being the shortest way to get what we need.